_source/_posts/2022-01-30-local-https-java.adoc

```diff
@@ -0,0 +1,278 @@
+
+Start Nginx with `docker run` mapping the above files as volumes to configure the service.
+
+====
+[source,bash]
+----
+docker run \
+    -v $(pwd)/nginx-default.conf.template:/etc/nginx/templates/default.conf.template \ # <.>
+    -v $(pwd)/devcerts:/etc/nginx/certs \ # <.>
+    -p 443:443 \
+    -p 80:80 \
+    --env TARGET_PORT=8080 \ # <.>
+    nginx
+----
+<.> The Nginx configuration file template.
+<.> The path to the certificates.
+<.> The port where your application is running on `localhost`.
+====
+
+NOTE: You can codify this solution using a https://github.com/bdemers/localhost-tls/blob/main/docker-compose.yml[docker-compose.yml file] if you prefer!
+
+Test it out! Open up your browser to `\https://localhost/`.
+
+== Configure Spring Boot to use HTTP
+
+The previous options used a separate service to handle TLS; this time, I'll configure Spring Boot's embedded Tomcat server to terminate TLS directly.
+
+Generate another set of certificates with `mkcert`, but this time set the output format to `pkcs12`, which is a format the Java Keystore understands.
+
+[source,bash]
+----
+# create the directory
+mkdir -p ~/.config/spring-boot
+
+# generate a certificate in pkcs12 format
+mkcert -pkcs12 -p12-file ~/.config/spring-boot/local-tls.p12 localhost
+----
+
+Create a properties file `~/.config/spring-boot/spring-boot-devtools.properties` containing the server port and newly genrated key store location.
+
+[source,properties]
+----
+# Set the port
+server.port=8443
+
+# configure the key store path
+server.ssl.key-store=${user.home}/.config/spring-boot/local-tls.p12
+----
+
+Now restart your Spring Boot application, and open your browser to `\https://localhost:8443` to access your newly secured application!
+
+[TIP]
+====
+I didn't need to make any changes to my application because it's using https://docs.spring.io/spring-boot/docs/current/reference/html/using.html[Spring Boot DevTools], which automatically includes the above properties file. If you are *not* using DevTools, you can mimic this behavior by setting an environment variable before starting your application:
+
+[source,bash]
+----
+SPRING_CONFIG_IMPORT='${user.home}/.config/spring-boot/spring-boot-devtools.properties'
+----
+====
+
+== Bonus: add a hostname alias
+
+The above examples use `localhost`, which works, but you might want to use a different hostname. One easy way to do this is to add a hostname alias to your `/etc/hosts` file:
+
+[source,bash]
+----
+sudo sh -c 'echo "127.0.0.1 local.example" >> /etc/hosts'
+----
+
+In the above examples, when generating certificates with `mkcert`, replace `localhost` with your domain name, for example:
+
+[source,bash]
+----
+mkcert -key-file devcerts/key.pem -cert-file devcerts/cert.pem local.example
+----
+
+WARNING: Use a domain name that you own or NOT in the official https://data.iana.org/TLD/tlds-alpha-by-domain.txt[top-level domain list] to ensure there are no conflicting or hijacked DNS entries.
+
```

Comment:
I'd like to see a section on the downsides of using TLS for local development. I was a big advocate for HTTPS locally a couple of years ago when I [wrote a post on it](https://developer.okta.com/blog/2019/03/07/spring-microservices-https-oauth2). However, I found myself reverting back to HTTP because I experienced caching issues. For example, Chrome would always redirect to HTTPS and I'd have to flush the cache to fix it.

[neverssl.com](http://neverssl.com) is a good tip for joining wi-fi when on the road too, but might not fit into this post.

Reply:
I love this idea! I try to add a few lines about it!